### ISSUE TITLE
split backup crashes signal

### ISSUE THREAD

**Issue by dethorpe** [bepaald/signalbackup-tools#127]
Being that my signal backup has grown over 10 GB I decided it would be best to divide it up by years and restore the current one (2023) to my phone.

signalbackup-tools_win.exe \\192.168.1.19\ruins\Backups\Signal\signal-2023-06-21-10-35-56.backup XXX --croptodates "2023-01-01 00:00:00","2024-01-01 00:00:00" --output C:\sig\2023.backup --passphrase XXX

It all seems to have gone fine but now signal crashes after running for 2 seconds. Not sure how to diagnose this further.

**Comment by bepaald** [comment]
I have to admit I don't believe the date-cropping option gets used a lot. I can imagine it has developed some small problems over the years, given all the many database changes. It should be possible to fix them, I will investigate.

But it would be extremely useful to get a debuglog or `adb logcat` output of Signal crashing, there is a good chance it will tell me where to look for the problem. So, if at all possible, I would love to see that.

**Comment by dethorpe** [comment]
I’m happy to help! However easier said than done. Android studio is not currently setup and I of course need to revert the db back to an unstable state, so please be patient. 

I’m surprised that this isn’t used a lot! I can’t be the only person dealing with too much backups. Am I missing a note obvious solution?

**Comment by bepaald** [comment]
>I’m happy to help! However easier said than done. Android studio is not currently setup and I, of course, need to revert the db back to an unstable state, so please be patient.

Of course, take your time. I'll still investigate the code on my end and I might try to reproduce, though my testing-backups are all fairly small with a limited number of contacts. This reduces the chance of running into the same bugs.

I don't think installing the entire Android Studio is required for `adb`, just the 'SDK Platform Tools' should be enough (not sure though, maybe it's different on Windows, but Android Studio is huge just to get that small tool).

>I’m surprised that this isn’t used a lot! I can’t be the only person dealing with too much backups. Am I missing a more obvious solution?

Well, of course, I don't actually know what functions people use, unless I get bug reports or thank-you-notes. Both of those are mostly about the merging, importing from desktop and exporting options.

If I needed to shrink my backup, I think I'd probably use the [deleting attachments option](https://github.com/bepaald/signalbackup-tools#deleting_attachments). Your backup may differ, in my experience the backup size is >99% attachment data, so by doing `--deleteattachments --onlyoldertthan "2023-01-01 00:00:00" -o [output]` you should have almost the same size reduction, but still have your full message history (minus the attachments). You could also replace the attachments with small placeholders (though the placeholders' sizes do add up).

_But_, my main suspect for the bug you are experiencing is a somewhat complex function that also happens to be called during the `deleteattachements` procedure, so you may run in to the exact same thing. Also, just like with the crop-function, I'm not sure this has been used since I developed it on request (I haven't had bug reports about it).

**Comment by dethorpe** [comment]
>the backup size is >99% attachment data
yeah, I've been on the fence about going this direction. I even remember finding a project a year back that just downscaled images to save space. I just can't stand the data loss! I'm such a data hoarder. Maybe... if we can't work this out I'll archive a full backup with everything and load up an attachment stripped version to the phone. But after a few cycles of that... the new achieves will have a long tail of redundant text. -_-

**Comment by bepaald** [comment]
Haha, I'm a bit of a data hoarder as well, that's a large part of why this program exists at all! But, I don't quite understand how deleting just the attachments causes more data loss than deleting the attachments _and_ the text content? Surely in either case you'll still keep the full unedited backup archived somewhere?

I guess the downscaling project was [this one](https://github.com/cycneuramus/signal-backup-shrink)? That actually uses this program too (see #68).

**Comment by dethorpe** [comment]
>I don't quite understand how deleting just the attachments causes more data loss than deleting the attachments and the text content?

Cuz I'm not. I'm just tossing the 20XX.backup file (with pics and text) onto my NAS where I will probably forget about it forever unless I ever want something on it and discover that I have insurmountable problems getting that since nobody else is doing partitioned backups and you've found other rewarding ways to spend your time =)

**Comment by dethorpe** [comment]
>--croptodates "2023-01-01 00:00:00","2024-01-01 00:00:00"

On no! Me thinks I've been foolish. Is the output of this the year 2023 or the backup without the year 2023?

**Comment by bepaald** [comment]
The output would be the year 2023. :-)

To get everything before 2023 you would do something like `--croptodates "0","2023-01-01 00:00:00"`.

**Comment by dethorpe** [comment]
Oh good. I was doing it right.

**Comment by LukeWheeldon** [comment]
Same issue here when simply running this command:
`signalbackup-tools [input] [passphrase] --deleteattachments --output [output] (--opassphrase [newpassphrase])`

I was able to solve the problem by following it with:
`signalbackups-tools [input] [passphrase] --output [outputfile] (--opassphrase [newpassphrase])`

Crossing fingers I haven't lost anything / much, due to the corruption.

Like you folks, I am a data hoarder; my method to make this sustainable is to export all medias to long term storage once every year. I run fdupes -dN to prune duplicates as necessary, and then prune all medias from the signal backup to finally import it into my Android device. It's a bit of work, but it's only once a year.

Hopefully this can help narrow down the culprit:

```
 *** Starting log: 2024-01-01 17:39:04 ***
signalbackup-tools (signalbackup-tools) source version 20240101.173827
BACKUPFILE VERSION: 1
BACKUPFILE SIZE: 68385873
COUNTER: 2450643038
Reading backup file...
FRAME 151467 (100.0%)...
Read entire backup file...
done!
Database version: 211

WARNING Foreign key constraint violated.
------------------------------------
| table       | parent      | fkid |
------------------------------------
| msl_message | msl_payload | 0    |
------------------------------------

WARNING Foreign key constraint violated.
------------------------------------
| table       | parent      | fkid |
------------------------------------
| msl_message | msl_payload | 0    |
------------------------------------
```

**Comment by bepaald** [comment]
Hi @LukeWheeldon, I'm not completely sure this is the same issue actually. The foreign key constraint violation will certainly cause a crash in Signal when importing the backup, but from what I can see from the output you posted, the violation is already in the backup file before any operations are performed.

Is this backup fresh from Signal (not already processed by this program)?

To fix the foreign key, you will probably want to 
```
signalbackup-tools [input] [passphrase] --runsqlquery "DELETE FROM msl_message WHERE payload_id NOT IN (SELECT _id FROM msl_payload)" --runsqlquery "pragma foreign_key_check"
```

This should hopefully show:
```
 * Executing query: DELETE FROM msl_message WHERE payload_id NOT IN (SELECT _id FROM msl_payload)
Modified [N] rows
 * Executing query: pragma foreign_key_check
(no results)
```

If it does, you can apply the changes by adding `--output [fixed.backup]` to the above command. If the output is different (it's possible the deletion causes another key constraint, or maybe I made a mistake in that query), please report back.

**Comment by LukeWheeldon** [comment]
Thanks for the quick help!

This is what I am getting on the original backup:
```
Database version: 211
 * Executing query: DELETE FROM msl_message WHERE payload_id NOT IN (SELECT _id FROM msl_payload)
Modified 0 rows
 * Executing query: pragma foreign_key_check
 ```

And what I get on the pruned file (from the command ran in my previous message:
```
Database version: 211

WARNING Foreign key constraint violated.
------------------------------------
| table       | parent      | fkid |
------------------------------------
| msl_message | msl_payload | 0    |
------------------------------------
 * Executing query: DELETE FROM msl_message WHERE payload_id NOT IN (SELECT _id FROM msl_payload)
Modified 239 rows
 * Executing query: pragma foreign_key_check
(no results)
```

**Comment by bepaald** [comment]
Ah, ok, so the key constraint violation only occurs after `--deleteattachments`, that is certainly a bug in this program then.

I do not immediately see where this problem comes from, but I'll take a closer look when I get back from work this afternoon. In the meantime, I wouldn't mind the full output from the original `signalbackup-tools [input] [passphrase] --deleteattachments --output [output] (--opassphrase [newpassphrase])` command if possible. Thanks!

> Attempting to load the pruned + pragma foreign_key_check backup results in a crash as well. Loading the original Signal backup is working fine.

So, just to be clear, the 'pragma foreign_key_check backup' is one that is saved by
```
signalbackup-tools [input] [passphrase] --runsqlquery "DELETE FROM msl_message WHERE payload_id NOT IN (SELECT _id FROM msl_payload)" --runsqlquery "pragma foreign_key_check" -o [NEW_FIXED_OUTPUT]
```
Just running the above without the `-o` option will not have modified the backup file. Does the above command  show 
```
 * Executing query: pragma foreign_key_check
(no results)
```
And then `NEW_FIXED_OUTPUT` still crashes Signal? In that case, something else is going on.

**Comment by LukeWheeldon** [comment]
Correct, I understand the -o is necessary to create a new file.

```
 *** Starting log: 2024-01-02 03:39:13 ***
signalbackup-tools (signalbackup-tools) source version 20240101.173827
BACKUPFILE VERSION: 1
BACKUPFILE SIZE: 3304446661
COUNTER: 2450643038
Reading backup file...
FRAME 160455 (100.0%)...
Read entire backup file...
done!
Database version: 211
Deleted: 2792 'part'-entries.
  Deleting unused attachments...
cleanDatabaseByMessages
  Deleting attachment entries from 'part' not belonging to remaining message entries
  Deleting other threads from 'thread'...
updateThreadsEntries
  Dealing with thread id: 1, 3, 4, 13, 14, 5, 9, 37, 63, 46, 22, 15, 31, 17, 41, 25, 55, 11, 20, 65, 30, 49, 21, 36, 45, 18, 27, 19, 28, 29, 6
, 7, 16, 8, 10, 12, 23, 24, 26, 43, 32, 38, 33, 34, 35, 62, 53, 42, 44, 47, 48, 54, 50, 51, 52, 56, 60, 57, 58, 59, 61, 68, 66, 67
  Deleting entries from 'mention' not belonging to remaining mms entries
  Deleting removed groups...
  Deleting unneeded MessageSendLog entries...
  Deleting reactions to non-existing messages
  Deleting call details from non-existing messages
  Deleting unreferenced recipient entries...
  Deleted 209 unreferenced recipients
  Deleted 1 unneeded distribution_lists
  Deleting unneeded notification profiles entries...
  Deleting unused avatars...
  Deleting unused attachments...
  Delete others from 'identities'
  Deleting group_receipts entries from deleted messages...
  Deleting group_receipts from non-existing recipients
  Deleting drafts from deleted threads...
  Deleting remapped recipients for non existing recipients
  Vacuuming database

WARNING Foreign key constraint violated.
------------------------------------
| table       | parent      | fkid |
------------------------------------
| msl_message | msl_payload | 0    |
------------------------------------

Exporting backup to 'test.backup'
Writing HeaderFrame...
Writing DatabaseVersionFrame...
Writing SqlStatementFrame(s)...
  Dealing with table 'avatar_picker'... 0/0 entries...done
  Dealing with table 'recipient'... 81/81 entries...done
  Dealing with table 'thread'... 64/64 entries...done
  Dealing with table 'message'... 92026/92026 entries...done
  Dealing with table 'call'... 113/113 entries...done
  Dealing with table 'call_link'... 0/0 entries...done
  Dealing with table 'cds'... 277/277 entries...done
  Dealing with table 'chat_colors'... 0/0 entries...done
  Dealing with table 'distribution_list'... 0/0 entries...done
  Dealing with table 'distribution_list_member'... 0/0 entries...done
  Dealing with table 'donation_receipt'... 4/4 entries...done
  Dealing with table 'drafts'... 0/0 entries...done
  Dealing with table 'emoji_search'... 0/0 entries...done
  Dealing with table 'groups'... 7/7 entries...done
  Dealing with table 'group_membership'... 49/49 entries...done
  Dealing with table 'group_receipts'... 37674/37674 entries...done
  Dealing with table 'identities'... 0/0 entries...done
  Dealing with table 'kyber_prekey'... 708/708 entries...done
  Dealing with table 'mention'... 48/48 entries...done
  Dealing with table 'msl_payload'... 1019/1019 entries...done
  Dealing with table 'msl_message'... 1310/1310 entries...done
  Dealing with table 'msl_recipient'... 2107/2107 entries...done
  Dealing with table 'notification_profile'... 0/0 entries...done
  Dealing with table 'notification_profile_allowed_members'... 0/0 entries...done
  Dealing with table 'notification_profile_schedule'... 0/0 entries...done
  Dealing with table 'part'... 0/0 entries...done
  Dealing with table 'payments'... 0/0 entries...done
  Dealing with table 'pending_pni_signature_message'... 0/0 entries...done
  Dealing with table 'pending_retry_receipts'... 0/0 entries...done
  Dealing with table 'reaction'... 15558/15558 entries...done
  Dealing with table 'remapped_recipients'... 3/3 entries...done
  Dealing with table 'remapped_threads'... 0/0 entries...done
  Dealing with table 'remote_megaphone'... 10/10 entries...done
  Dealing with table 'sender_key_shared'... 0/0 entries...done
  Dealing with table 'sender_keys'... 0/0 entries...done
  Dealing with table 'sticker'... 103/103 entries...done
  Dealing with table 'storage_key'... 4/4 entries...done
  Dealing with table 'story_sends'... 0/0 entries...done
Writing SharedPrefFrame(s)...
Writing KeyValueFrame(s)...
Writing Avatars...
Writing EndFrame...
Done! Wrote 68385873 bytes.
```

I'm going to have to re-load the original backup, which may make it difficult to troubleshoot further. I have another phone, maybe I could use that to do some tests. I won't be able to do phone tests until next week though. I do have access to the backups as well as the linux machine where signalbackup-tools is installed.

Thanks.

**Comment by bepaald** [comment]
>Correct, I understand the -o is necessary to create a new file.

Ok, sorry, just making sure :)

So, that to me means you are running in to two bugs! 

I was really losing my mind over the foreign key-problem. Turns out there are triggers on the `msl_payload` table to delete entries when messages or attachments are deleted. The corresponding `msl_message` entries are not deleted however, because they are deleted by a foreign key constrains (`ON DELETE CASCADE`) instead of a trigger, which are not enforced by default in sqlite. This actually causes a bit of a circular dependency type problem in this case. Either way, it should be fixed now. You could give it a try (no need to restore on a phone for this, just assert that no foreign key constraint error is shown when running with `--deleteattachments`). Thanks for reporting!

As for the second one: I did fix a logic error in `deleteAttachments()`, though I don't think it would cause a crash in Signal upon restoring (but I could be wrong). I'm afraid for the second bug (if still present), you would need to restore to a phone, and have it connected to a computer with `adb logcat` running to catch the crash (maybe a debug log from Signal would work as well).

Thanks!

**Comment by LukeWheeldon** [comment]
Looking better indeed! I have tried restoring in Signal on a different phone and it did not crash. Thanks a lot @bepaald.

```
 *** Starting log: 2024-01-03 23:53:31 ***                                                                                            [48/768]
signalbackup-tools (signalbackup-tools) source version 20240103.201232
BACKUPFILE VERSION: 1
BACKUPFILE SIZE: 3304446661
COUNTER: 2450643038
Reading backup file...
FRAME 160455 (100.0%)...
Read entire backup file...
done!
Database version: 211
Deleted: 2792 'part'-entries.
Deleted 937 empty messages
  Deleting unused attachments...
cleanDatabaseByMessages
  Deleting attachment entries from 'part' not belonging to remaining message entries
  Deleting other threads from 'thread'...
updateThreadsEntries
  Dealing with thread id: 1, 3, 4, 13, 14, 5, 9, 37, 63, 46, 22, 15, 31, 17, 41, 25, 55, 11, 20, 65, 30, 49, 21, 36, 45, 18, 27, 19, 28, 29, 6
, 7, 16, 8, 10, 12, 23, 24, 26, 43, 32, 38, 33, 34, 35, 62, 53, 42, 44, 47, 48, 54, 50, 51, 52, 56, 60, 57, 58, 59, 61, 68, 66, 67
  Deleting entries from 'mention' not belonging to remaining mms entries
  Deleting removed groups...
  Deleting unneeded MessageSendLog entries... (429, 8, 748)
  Deleting reactions to non-existing messages... (538)
  Deleting call details from non-existing messages... (0)
  Deleting unreferenced recipient entries...
  Deleted 209 unreferenced recipients
  Deleted 1 unneeded distribution_lists
  Deleting unneeded notification profiles entries...
  Deleting unused avatars...
  Deleting unused attachments...
  Delete others from 'identities'
  Deleting group_receipts entries from deleted messages...
  Deleting group_receipts from non-existing recipients
  Deleting drafts from deleted threads...
  Deleting remapped recipients for non existing recipients
  Vacuuming database

Exporting backup to 'signal-2024-01-02-08-20-36-textonly.backup'                                                                      [11/768]
Writing HeaderFrame...
Writing DatabaseVersionFrame...
Writing SqlStatementFrame(s)...
  Dealing with table 'avatar_picker'... 0/0 entries...done
  Dealing with table 'recipient'... 81/81 entries...done
  Dealing with table 'thread'... 64/64 entries...done
  Dealing with table 'message'... 92026/92026 entries...done
  Dealing with table 'call'... 113/113 entries...done
  Dealing with table 'call_link'... 0/0 entries...done
  Dealing with table 'cds'... 277/277 entries...done
  Dealing with table 'chat_colors'... 0/0 entries...done
  Dealing with table 'distribution_list'... 0/0 entries...done
  Dealing with table 'distribution_list_member'... 0/0 entries...done
  Dealing with table 'donation_receipt'... 4/4 entries...done
  Dealing with table 'drafts'... 0/0 entries...done
  Dealing with table 'emoji_search'... 0/0 entries...done
  Dealing with table 'groups'... 7/7 entries...done
  Dealing with table 'group_membership'... 49/49 entries...done
  Dealing with table 'group_receipts'... 37674/37674 entries...done
  Dealing with table 'identities'... 0/0 entries...done
  Dealing with table 'kyber_prekey'... 708/708 entries...done
  Dealing with table 'mention'... 48/48 entries...done
  Dealing with table 'msl_payload'... 1019/1019 entries...done
  Dealing with table 'msl_message'... 1071/1071 entries...done
  Dealing with table 'msl_recipient'... 2107/2107 entries...done
  Dealing with table 'notification_profile'... 0/0 entries...done
  Dealing with table 'notification_profile_allowed_members'... 0/0 entries...done
  Dealing with table 'notification_profile_schedule'... 0/0 entries...done
  Dealing with table 'part'... 0/0 entries...done
  Dealing with table 'payments'... 0/0 entries...done
  Dealing with table 'pending_pni_signature_message'... 0/0 entries...done
  Dealing with table 'pending_retry_receipts'... 0/0 entries...done
  Dealing with table 'reaction'... 15558/15558 entries...done
  Dealing with table 'remapped_recipients'... 3/3 entries...done
  Dealing with table 'remapped_threads'... 0/0 entries...done
  Dealing with table 'remote_megaphone'... 10/10 entries...done
  Dealing with table 'sender_key_shared'... 0/0 entries...done
  Dealing with table 'sender_keys'... 0/0 entries...done
  Dealing with table 'sticker'... 103/103 entries...done
  Dealing with table 'storage_key'... 4/4 entries...done
  Dealing with table 'story_sends'... 0/0 entries...done
Writing SharedPrefFrame(s)...
Writing KeyValueFrame(s)...
Writing Avatars...
Writing EndFrame...
Done! Wrote 68368187 bytes.
```

**Comment by bepaald** [comment]
Thanks for taking the time to test en report back. I think I can consider the problems you reported are fixed, so that's great.

But whether one of those was the same problem as originally reported here, I'm still not sure. @dethorpe: is this still an active issue for you? Any chance of re-testing or getting a debug log from the crash? I'd still love to investigate this further, but if the issue is gone or circumstances prohibit you from reproducing or getting more info on it (which is fine of course), this issue might as well be closed for now. If it comes up again, a new issue can of course always be opened.

Thanks!

**Comment by bepaald** [comment]
Just closing this, as one of you had the issues solved, and the other is MIA. This can always be reopened of course (or a new issue can be created).

Thanks!
